dbms/src/Flash/Mpp/HashPartitionWriter.cpp

```diff
@@ -14,66 +14,189 @@
 {
     rows_in_blocks = 0;
     partition_num = writer_->getPartitionNum();
     RUNTIME_CHECK(partition_num > 0);
     RUNTIME_CHECK(dag_context.encode_type == tipb::EncodeType::TypeCHBlock);
+
+    switch (data_codec_version)
+    {
+    case MPPDataPacketV0:
+        break;
+    case MPPDataPacketV1:
+    default:
+    {
+        if (batch_send_min_limit_mem_size < 0)
+        {
+            // set upper limit if not specified
+            batch_send_min_limit_mem_size = std::min(MAX_BATCH_SEND_MIN_LIMIT_MEM_SIZE, 1024 * 64 * partition_num /* 64KB * partition-num */);
+        }
+        for (const auto & field_type : dag_context.result_field_types)
+        {
+            expected_types.emplace_back(getDataTypeByFieldTypeForComputingLayer(field_type));
+        }
+        break;
+    }
+    }
 }
 
 template <class ExchangeWriterPtr>
 void HashPartitionWriter<ExchangeWriterPtr>::flush()
 {
-    if (rows_in_blocks > 0)
+    if (0 == rows_in_blocks)
+        return;
+
+    switch (data_codec_version)
+    {
+    case MPPDataPacketV0:
+    {
         partitionAndWriteBlocks();
+        break;
+    }
+    case MPPDataPacketV1:
+    default:
+    {
+        partitionAndWriteBlocksV1();
+        break;
+    }
+    }
 }
 
 template <class ExchangeWriterPtr>
-void HashPartitionWriter<ExchangeWriterPtr>::write(const Block & block)
+void HashPartitionWriter<ExchangeWriterPtr>::writeImplV1(const Block & block)
 {
-    RUNTIME_CHECK_MSG(
-        block.columns() == dag_context.result_field_types.size(),
-        "Output column size mismatch with field type size");
     size_t rows = block.rows();
     if (rows > 0)
     {
         rows_in_blocks += rows;
+        mem_size_in_blocks += block.bytes();
         blocks.push_back(block);
     }
+    if (mem_size_in_blocks > batch_send_min_limit_mem_size)
+        partitionAndWriteBlocksV1();
+}
 
+template <class ExchangeWriterPtr>
+void HashPartitionWriter<ExchangeWriterPtr>::writeImpl(const Block & block)
+{
+    size_t rows = block.rows();
+    if (rows > 0)
+    {
+        rows_in_blocks += rows;
+        blocks.push_back(block);
+    }
     if (static_cast<Int64>(rows_in_blocks) > batch_send_min_limit)
         partitionAndWriteBlocks();
 }
 
+template <class ExchangeWriterPtr>
+void HashPartitionWriter<ExchangeWriterPtr>::write(const Block & block)
+{
+    RUNTIME_CHECK_MSG(
+        block.columns() == dag_context.result_field_types.size(),
+        "Output column size mismatch with field type size");
+
+    switch (data_codec_version)
+    {
+    case MPPDataPacketV0:
+    {
+        return writeImpl(block);
+    }
+    case MPPDataPacketV1:
+    default:
+    {
+        return writeImplV1(block);
+    }
+    }
+}
+
+template <class ExchangeWriterPtr>
+void HashPartitionWriter<ExchangeWriterPtr>::partitionAndWriteBlocksV1()
+{
+    assert(rows_in_blocks > 0);
+    assert(mem_size_in_blocks > 0);
+
+    if unlikely (blocks.empty())
```

Comment:
since there is `assert(rows_in_blocks > 0)`, I think blocks should never be empty?

Reply:
Fixed